### ISSUE TITLE
Tekton: v1beta1 fields in v1alpha1 model

### ISSUE THREAD

**Issue by Fabian-K** [fabric8io/kubernetes-client#2159]
Hi,

there is an issue with the parallel support of v1alpha1 and v1beta1 for tekton.

Let´s use the Task as an example as this contains a breaking change from v1alpha1 to v1beta1 with respect to params.

v1alpha1 supplies params like the following:
````
spec:
  inputs:
    params:
      - name: someURL
        type: string
````

v1beta1, however, expects this:
````
spec:
  params:
    - name: someURL
      type: string
````

The java generator generates for v1alpha1 both: the inputs field as well as the params field. This reflects the source code, as according to the go definition, the v1alpha1 TaskSpec is the v1beta1 TaskSpec plus the inputs and outputs field.
https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1alpha1/task_types.go#L46-L58

During the serialization process for a v1alpha1 Task, the v1beta1 fields are included as an empty array.
````
{
  "apiVersion": "tekton.dev/v1alpha1",
  "kind": "Task",
  "spec": {
    "inputs": {
      "params": [
        {
          "name": "param",
          "type": "string"
        }
      ]
    },
    "params": []
  }
}
````

This results in a rejection by tektons admission webhook when using a tekton version that only supports v1alpha1:
````
Message: admission webhook "webhook.tekton.dev" denied the request: mutation failed: cannot decode incoming new object: json: unknown field "params". Received status: Status(apiVersion=v1, code=400, details=null, kind=Status, message=admission webhook "webhook.tekton.dev" denied the request: mutation failed: cannot decode incoming new object: json: unknown field "params", metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=BadRequest, status=Failure, additionalProperties={})
````

I assume this works when using tekton that supports both v1alpha1 and v1beta1 but I have not checked this.

As a result, the intention of supporting v1alpha1 and v1beta1 in parallel for backward compatibility does not work 😢

I see 2 possibilities to fix this:
1) the fields in the go model are marked with omitempty. We could enhance the generator to respect this and make sure that during serialization this is not included.
2) we create 2 schemas:
  - one containing only v1alpha1 using the tekton-pipelines dependency with version 0.10.2 
  - one containing only v1beta1 using the tekton-pipelines dependency with version 0.11.2 

I would prefer 2) as this results in a better java model. With 1) we might achieve backward compatibility, however the java API for v1alpha1 would also include fields of v1beta1 and we would have to rely on endusers only setting the right fields

What do you think about this?

Best regards,
Fabian

@rohanKanojia notifying you as IMO this should be fixed before a next release

**Comment by rohanKanojia** [comment]
Hmm, :thinking: we're doing something similar in case of kubernetes model annotator for omitting non empty things:
https://github.com/fabric8io/kubernetes-client/blob/1150135f9febfd000396c031532efe5856037043/kubernetes-model/kubernetes-model-annotator/src/main/java/io/fabric8/kubernetes/annotator/KubernetesTypeAnnotator.java#L134-L139

**Comment by Fabian-K** [comment]
I see, I think I should definitely update the generator (& common annotator added in https://github.com/fabric8io/kubernetes-client/pull/2157) to support this as well. As of now, this is missing.

Even if this fixes the backward compatibility, generating separate schemas (based on different go dependency versions) still makes sense for me as the java API is just much better

**Comment by rohanKanojia** [comment]
@Fabian-K : Are you saying we should have two different jars for `v1alpha1` and `v1beta1` model(each model being generated from a separate schema?

**Comment by Fabian-K** [comment]
Not necessarily to different jars (we could also achieve it using different namespaces) but sure, why not split it into different jars :)

Important is to use an older version of the tekton go dependency when generating v1alpha1
